### PR TITLE
Alternative implementation for #1368 using singleton.

### PR DESCRIPTION
**Pull request by ams-tschoening**
PR #1369 for #1368 didn't work upstream and make sense to not work, because `QApplication` is needed too early in the process of starting up the app.

So I suggest an alternative implementation using a singleton which is created once AFTER the app started up and can then use `QApplication` to retrieve data as necessary. Changing the formerly used namespace to a class implementing a singleton requires the least amount of changes and allows more control over what gets created when how, especially with an instance of `QApplication` being available.

Don't get confused by the many changes: What has formerly been a constant in a namespace and got a value set, is now simply changed to an instance variable of some class and gets a value set as well in the member initialization of the CTOR of the class. Syntax is very similar, it's only `SomeVar = SomeVal` vs. `SomeVar(SomeVal)` mostly. Most other changes are using the new class instead of the namespace: `GlobalAttributes::SomeVar` vs. `GlobalAttributes::get().SomeVar`.

The singleton is only used as PoC, that can be reconsidered and I don't care much about the number of instances of `GlobalAttributes` being available. The important thing is in my opinion switching `GlobalAttributes` to a class so that one has more control over when those attributes get created.

While things compile successfully and using Process Monitor it seems that some necessary paths are read successfully this way, the app still crashes:

> Exception code=0xc00000fd flags=0x0 at 0x000000006E5C3BC4

Don't know how to debug this further easily currently and would like to discuss if this is an approach @rkhaotix wants to follow at all or not.

### CODE REVIEW

**Comment by ams-tschoening** [comment]
The plugins need to be adopted as well, but things are easy:

[dummy.cpp.diff.txt](https://github.com/pgmodeler/pgmodeler/files/4026628/dummy.cpp.diff.txt)
[xml2object.cpp.diff.txt](https://github.com/pgmodeler/pgmodeler/files/4026629/xml2object.cpp.diff.txt)
[xml2objectwidget.cpp.diff.txt](https://github.com/pgmodeler/pgmodeler/files/4026630/xml2objectwidget.cpp.diff.txt)

**Comment by rkhaotix** [comment]
Honestly, I don't like any singleton approach. :(
I had lots of problem with singletons (mainly related to widgets) in the earlier versions of pgModeler and decided to remove them once for all from the project. Let me try to think on something too... Currently I'm working on porting the code to Qt 5.14 and make it buildable in Qt 5.12 and earlier versions but I'll come up with a proper fix for this issue.

**Comment by ams-tschoening** [comment]
I think the important thing is to switch to proper classes and instances instead of the current namespace-approach, so that you can control when and how those instances are created. That's what may changes are showing only in the end, how many instances are available etc. is less important. The singleton is only what was nearest to what you have used before and things still read very similar etc.

**Comment by rkhaotix** [comment]
I understand and appreciate your desire to make things work properly, but I fear that your PR is not suitable for the problem, yet. Let me try to explain with a bit more details...

First, as I understand, your singleton implementation tries to resolve the folders inside pgModeler installation path which, in a certain way, is really correct. But you're supposing that the user (or a developer or package maintainer) will always install pgModeler that way (say ```c:\program files\pgmodeler\[schemas, lang, etc, etc]``` or in a similar way on the other platforms). But imagine the following paths configuration after running ```qmake -r pgmodeler.pro PREFIX+=/usr CONFDIR+=/etc/pgmodeler/conf PLUGINSDIR+=/usr/lib/plugins``` (this example can be easily reproducible on Windows which seems to be your main OS):

```
  ** pgModeler build details ** 
 
  PREFIX        = /usr 
  BINDIR        = /usr/bin 
  PRIVATEBINDIR = /usr/lib/pgmodeler/bin 
  PRIVATELIBDIR = /usr/lib/pgmodeler 
  PLUGINSDIR    = /usr/lib/plugins 
  SHAREDIR      = /usr/share/pgmodeler 
  CONFDIR       = /etc/pgmodeler/conf 
  DOCDIR        = /usr/share/pgmodeler 
  LANGDIR       = /usr/share/pgmodeler/lang 
  SAMPLESDIR    = /usr/share/pgmodeler/samples 
  SCHEMASDIR    = /usr/share/pgmodeler/schemas 
 
* To change a variable value run qmake again setting the desired value e.g.: 
  > qmake PREFIX+=/usr/local -r pgmodeler.pro 
 
* Proceed with build process by running: 
  >  make && make install 
```

Unless you tweak the ```PGMODELER_???``` environment variables, your implementation will try to look for that folders in ```/usr/bin``` (as the fallback result of ```getPathFromEnv()```) because is where ```pgmodeler``` executable is. So, for a "scattered" installation like the example above your solution and to original one made by me aren't the ideal ones. This is where the environment variables can be used to minimize the problems.

Honestly, I still don't know how to capture the correct paths of that individual folders during runtime (considering that they can be even more scattered compared to what I demonstrated). What I know is look for that folders only inside the executable's path is not the right/permanent solution. So, do you understand why this is a bit complex to implement?

Sorry if my English doesn't help to clarify things! :(

**Comment by rkhaotix** [comment]
It's just came up to me that a way to solve that is a combination of a config file and a small tweak on ```getPathFromEnv()```.

Say we have a file named ```paths.conf``` which holds that set of paths mentioned above. Anytime pgModeler starts up it'll lookup the values of the paths from that file. If path configuration for a individual folder is not there, pgModeler then tries to look into the enviroment variables (as it does now) and finally if no environment variable is found it will then use the approach to look for that folder in the same directory as the executable. 

This is similar to the file ```qtplugins.conf``` which is used by Qt when starting up an application and configure its internal paths like plugins and other submodules paths.

**Comment by ams-tschoening** [comment]
> But you're supposing that the user (or a developer or package maintainer) will always install pgModeler that way (say c:\program files\pgmodeler\[schemas, lang, etc, etc] or in a similar way on the other platforms).

I'm not, but what you describe is simply the default and recommended for Windows. Doing it otherwise is an absolute minor corner case that could almost be ignored. But we even don't ignore it, but instead people can provide environment variables to implement that absolutely rare corner case.

On non-Windows, scattering things over different places is more common, especially on Linux. But that is already supported and will still be supported, because my changes don't change that at all. But even there using the installation dir as LAST FALLBACK when nothing else is available is the better choice than using the current working directory. Especially on Linux `$PATH` not even contains the current working directory for security reasons, yet you want to load plugins/code relative to that as fallback. That doesn't make sense and violates platform standards.

> Unless you tweak the PGMODELER_??? environment variables, your implementation will try to look for that folders in /usr/bin (as the fallback result of getPathFromEnv()) because is where pgmodeler executable is.

And that's the correct thing to do for a LAST FALLBACK when nobody cared about anything else.

> So, for a "scattered" installation like the example above your solution and to original one made by me aren't the ideal ones. This is where the environment variables can be used to minimize the problems.

But we are talking about the case when those environment variables are not available only. That's very unlikely on Linux, you are even distributing a custom start-script to circumvent that, but it's very likely e.g. on Windows, where your current installer fails to set those environment variables and things like file associations don't work properly anymore. That could be prevented by using a different fallback easily.

> Say we have a file named paths.conf which holds that set of paths mentioned above.

That doesn't help anything: How do you know where that file is stored? You either assume the current working dir and things will simply fail again like they are doing already, or will use the application installation dir and therefore am implementing exactly the fallback I'm proposing, only in far more difficult. Who writes that file on e.g. Windows? Your current installer doesn't seem to create the environment variables properly already, so why adding even more code for that file which needs to be maintained and stuff? Additionally, especially on Windows it's pretty common to simply relocate installation directories, which fails with the paths used in your new config file as well, but simply works when changing the fallback to the installation dir.

Think of it, this is the wrong approach and doesn't make anything easier. Everything breaks down to the question what a reasonable hard-coded fallback for `getPathFromEnv` is in case no environment variable, config file etc. can be found: The current directory or the "installation directory" and its simply the latter.

**Comment by rkhaotix** [comment]
> That doesn't help anything: How do you know where that file is stored? You either assume the current working dir and things will simply fail again like they are doing already, or will use the application installation dir and therefore am implementing exactly the fallback I'm proposing, only in far more difficult. Who writes that file on e.g. Windows? Your current installer doesn't seem to create the environment variables properly already, so why adding even more code for that file which needs to be maintained and stuff? Additionally, especially on Windows it's pretty common to simply relocate installation directories, which fails with the paths used in your new config file as well, but simply works when changing the fallback to the installation dir.

Yeah, after writting my previous comment and taking some fresh air I realized that the paths.conf is useless too... 

What makes me wonder is, despite the CWD fallback failing, what's wrong with environment variables usage? On windows I was using them when packaging the installation using InnoSetup. Since I moved to Qt installer framework in order to unify the installation maintenance I did not implemented the env. vars registration because of other problems with the development of 0.9.1 and 0.9.2.

But now that the stable release is here I can rework the installer to write those variables and do the proper file associations (as well that problem with start menu you mentioned). 

> That's very unlikely on Linux, you are even distributing a custom start-script to circumvent that...

I use a start-script on Linux just to avoid some inexperienced users to messing up with system's environment variables and break things. But we could easily eliminate that script and make the installer to configure the env.  vars. It was just a prudent solution to avoid problems to the users. Or, on the other hand, we could use the same approach (start-script) on Windows (I'm not the DOS batch script guy but I can do my best). 

What I want to make clear is that while I'm glad to see that you're involved with the solution of the problem I still want to see more options to solve the issue... The singleton one, again and honestly, is my last choice. Don't get me wrong, please.

**Comment by rkhaotix** [comment]
@ams-tschoening Reading again your PR, mainly on globalattributes.(h/cpp) I think we can use part of that implementation. I explain...

You created a a class named GlobalAttributes inside the namespace GlobalAttributes. Basically the class has all static attributes of the GlobalAttributes NS. What if we reduce that class to the set of attributes we need to solve the problem? The discussion here is about the default paths, right? What about we create a private class in GlobalAttributes named, for example, Paths or PgModelerPaths, and that class controls only that portion of things we're discussing and not involve the rest of the attributes of GlobalAttributes? Just to separate what need to be handled with app instance from those that not... I don't know if its possible, just a guess... But if you want to give it a try and have success please make a PR on "qt5.14-deprecation-fixes" so I can test with all changes I'm doing so far.

**Comment by rkhaotix** [comment]
Oh, forget it. GlobalAttributes IS A CLASS and not a child of namespace. Yeah, I'm gonna try it later .

**Comment by Maxzor** [comment]
c341c7ec does not compile, a syntax issue with GlobalAttributes namespace. (I don't know that declaration syntax so cannot fix it).

```
src/globalattributes.cpp:172:53: error: expected ‘{’ before ‘;’ token
  172 |   app.applicationDirPath() + QString("/pgmodeler")));
      |                                                     ^
src/globalattributes.cpp: At global scope:
src/globalattributes.cpp:195:1: error: expected unqualified-id before ‘{’ token
  195 | {

```

**Comment by ams-tschoening** [comment]
> What makes me wonder is, despite the CWD fallback failing, what's wrong with environment variables usage?

They are difficult to manage: Someone needs to set them, needs to set them for the correct user and stuff like that. You did that in InnoSetup for Windows, but forget/left out by purpose/... for the new QTIF-based installer and things broke. Even if the new installer had them in, because of the `AllUsers=true`-thing the new installer would only have set them for the installing user, not for that executing pgModeler in the end and things break as well in those setups. And I'm using those setups, installing as `Administrator` and running things as `tschoening`.

So what I propose is simply a change which improves chances that things still work without those environment variables. Don't get me wrong: I don't want to remove those environment variables entirely and didn't in this PR! I only want another fallback which works better than the current directory in most cases where those environment variables are not available for some reason.

> Or, on the other hand, we could use the same approach (start-script) on Windows (I'm not the DOS batch script guy but I can do my best).

I've already created such a script:

[pgModeler.cmd.txt](https://github.com/pgmodeler/pgmodeler/files/4029598/pgModeler.cmd.txt)

But the point is, that this script is simply doing what my PR does: Resolve paths based on the application directory instead of the current working directory. Additionally, that script can't be easily associated with files, because it doesn't provide file icons, it needs to be deploeyt with the application and stuff like that.

So simply changing the fallbacks in `GlobalAttributes` is the easier approach and instantly available on all platforms, you don't need to deploy additional files etc. Consider not only Linux vs. Windows, macOS prefers combining an application in one directory as well and even the results of your current build are a working application in one directory. So it simply makes sense to use that directory to resolve paths as fallback.

> Just to separate what need to be handled with app instance from those that not...

Things don't get much easier that way in my opinion. Currently you need an app instance for paths only, but I guarantee you that this will change for other settings in future as well, so creating all global attributes with an app instance is the easier approach in my opinion.

> c341c7e does not compile, a syntax issue. (I don't know that declaration syntax so cannot fix it).

Either remove the `;` or if that still fails use a `,`. I only tested on Windows yet and didn't get into that `#if defined`.

File: libutils/src/globalattributes.cpp [hunk 1]
```diff
@@ -20,156 +20,205 @@
...
+
+	fi.setFile(fallback_val);
+	return(fi.absoluteFilePath());
+}
+
+GlobalAttributes& GlobalAttributes::create(const QApplication& app)
+{
+	GlobalAttributes::get()._instance = new GlobalAttributes(app);
+
+	return *GlobalAttributes::get()._instance;
```
Comment (Maxzor) [inline libutils/src/globalattributes.cpp]: This is why it crashes. It goes in an endless loop.
Should be :
return *GlobalAttributes::get()._instance;

File: libutils/src/globalattributes.cpp [hunk 2]
```diff
@@ -20,156 +20,205 @@
...
+}
+
+const GlobalAttributes& GlobalAttributes::get()
+{
+	return *GlobalAttributes::get()._instance;
```
Comment (Maxzor) [inline libutils/src/globalattributes.cpp]: This is why it crashes. It goes in an endless loop.
Should be :
return *GlobalAttributes::_instance;

File: windeploy.sh [hunk 3]
```diff
@@ -3,7 +3,7 @@
 LOG=windeploy.log
 
 # Detecting current pgModeler version
-DEPLOY_VER=`cat libutils/src/globalattributes.cpp | grep PgModelerVersion | sed 's/PgModelerVersion=QString("//g' | sed 's/")//g'`
+DEPLOY_VER=`cat libutils/src/globalattributes.cpp | grep PgModelerVersion | sed 's/PgModelerVersion(QString("//g' | sed 's/")//g'`
```
Comment (Maxzor) [inline windeploy.sh]: ?

Comment (Maxzor) [reply windeploy.sh]: Now it compiles, and run, still don't really understand the purpose of this, and how the fallback dir changes.

**Comment by Maxzor** [comment]
Should be :
```
Application::Application(int &argc, char **argv) : QApplication(argc,argv)
{
	QTranslator *main_translator=nullptr, *plugin_translator=nullptr;
	GlobalAttributes::create(*this);
``` 
and 
```
GlobalAttributes& GlobalAttributes::create(const QApplication& app)
{
	GlobalAttributes::_instance = new GlobalAttributes(app);
}

const GlobalAttributes& GlobalAttributes::get()
{
	return *GlobalAttributes::_instance;
}
```

File: windeploy.sh (hunk 3 above)
Comment (ams-tschoening) [reply windeploy.sh]: `windeploy.sh` prints the current version of pgModeler on the shell and gets that value from `globalattributes.cpp` With switching to a class, that version number simply needs to be read slightly different.

**Comment by ams-tschoening** [comment]
> Shouldn't it be :

Yes, I've replaced usages of `GlobalAttributes` automatically on all files and simply forgot to remove those important parts. Thanks for recognizing, I thought things were more difficult and stopped debugging.

> Now it compiles, and runs, still don't really understand how the singleton helps in the fallback dir change ?

The fallback-paths were relative to the current working directory of the process before, but that's unreliable and breaks under various conditions. I changed that to use the application path provided by QT, but that seems only to be available after an instance of `QApplication` has been created. `GlobalAttributes` was a namespace containing constants before and couldn't access instances of `QApplication`, so I changed things to be a class which can be created at will with providing `QApplication`.

That's the important part of the story, storing only one instance or many of `GlobalAttributes` is not that important. Creating one instance only simply worked easiest as PoC.

**Comment by Maxzor** [comment]
Until the project enable cmake support in qt6, visualstudio builds are not a thing, and I could not get qtcreator to debug with msys2 qt. So i'd suggest to debug in a linux vm, since there is no proper windows debugging AFAIK.

**Comment by Maxzor** [comment]
The fix over 7b081f0 : http://ix.io/26G7

**Comment by Maxzor** [comment]
Related to your proposal, @rkhaotix recent adaptation leading to commit 9f9f006, and the general problem of environment variables vs runtime info, what do you think of global [QSettings](https://doc.qt.io/qt-5/qsettings.html#details) ?
I am working on [file selectors](https://github.com/pgmodeler/pgmodeler/issues/1321) and thinking of it.

**Comment by ams-tschoening** [comment]
Please be more detailed with your question and lets discuss this in #1321 then. Doesn't seem to fit here very well, but I'm not sure I understand what exactly your are asking me about. Sounds like which paths to store in QSettings, if at all or something completely different?

**Comment by Maxzor** [comment]
QSettings look to me like a solution that could solve a ton of problems :
- [application paths and fallbacks](https://doc.qt.io/qt-5/qsettings.html#fallback-mechanism)  the one we are trying to resolve here
- do properly [config files](https://doc.qt.io/qt-5/qsettings.html#platform-specific-notes)
- store/load [model tree states](https://github.com/pgmodeler/pgmodeler/blob/ea971caab2912542633bba1aa9f39faf24a4f0b8/libpgmodeler_ui/src/mainwindow.cpp#L1165), and basically any application state, wether UI or logic.
- store a handful of directories and default file paths : e.g. the folder where you usually store .dbm files, the folder where your csv data is, etc. which is #1321

I agree that this would need a separate discussion.
What I feel now is that QSettings class did probably not exists when @rkhaotix wrote the core of the application, so Raphaël basically wrote its own qsettings implementation which is roughly what GlobalAttributes + other dissemintaed features is. In the same way there was no Qt Undo/Redo stack so pgmodeler got its own class, but now Qt provides it.  There is room I believe for a consequent rewrite of GlobalAttributes, but is the added value enough, and cost acceptable, I do wonder.

**Comment by ams-tschoening** [comment]
> - [application paths and fallbacks](https://doc.qt.io/qt-5/qsettings.html#fallback-mechanism)  the one we are trying to resolve here

Don't think so, your link is about fallbacks `QSettings` use on its own to find its own managed configs. I don'T see how this can be used to decide at runtime where `pgmodeler.exe` is actually stored to know where to look for it schema files based on that path and stuff. Look at the platform specific notes:

https://doc.qt.io/qt-5/qsettings.html#platform-specific-notes

The paths mentioned there are only to manage settings of some app following some convention, it doesn't tell anything about where the app is actually stored. So that will be two different things and pgModeler still needs some way to decide for itself how to resolve some paths at runtime.

Besides that, your argument makes sense to me as well, `QSettings` is probably worth looking more detailed at.

**Comment by Maxzor** [comment]
> Don't think so, your link is about fallbacks `QSettings` use on its own to find its own managed configs. I don'T see how this can be used to decide at runtime where `pgmodeler.exe` is actually stored to know where to look for it schema files based on that path and stuff. Look at the platform specific notes:

You're right, I guess QSettings could be overriden by PgModeler for the particular BIN_DIR, SCHEMA_DIR etc., would need to think more about it.

**Comment by ams-tschoening** [comment]
Closing here, as #1368 has been closed by https://github.com/pgmodeler/pgmodeler/commit/af5a3e874e3988f505773a7b4147c18d784787d4 already.

**Pull request closed without merging**